File: src/client/gameFlow.cpp

#include "gameFlow.h"

Game::Game(Board *board_,
           Logic *logic_,
           Display *display_,
           Player *first,
           Player *second) : board_(board_),
                             logic_(logic_),
                             display_(display_) {
  players_[0] = first;
  players_[1] = second;
  validTurns_ = 0;
}

Game::~Game() {
  delete board_;
  delete display_;
  delete logic_;
  delete players_[0];
  delete players_[1];
}

void Game::run() {
  display_->printBoard(board_);
  int current = 0;
  // Game loop to play one turn until shouldStop() turns false.
  while (!shouldStop()) {
    playOneTurn(current);
    current = 1 - current;
  }

  cout << "The final score is: " << endl;
  for (int i = 0; i < NUM_OF_PLAYERS; i++) {
    cout << players_[i]->getSymbol() << " with "
         << players_[i]->getScore() << " points." << endl;
  }
  cout << endl;
  if (logic_->isTie(*players_[0], *players_[1])) {
    cout << "It's a tie!" << endl;
  } else {
    Player *winner = logic_->getWinner(players_);
    cout << "Winner is: '" << *winner << "' with score of "
         << winner->getScore()
         << endl;
  }
}

void Game::playOneTurn(int current) {
  cout << "current player: " << players_[current]->getSymbol() << endl;

  int numOfNoMoves =
      players_[current]->makeMove(*board_, *logic_, *display_);
  if (numOfNoMoves == 0) {
    validTurns_ = 0;
    return;
  }
  validTurns_ += numOfNoMoves;

}

bool Game::shouldStop() {
  return board_->isFull() || (validTurns_ >= NUM_OF_PLAYERS);
}

File: Console/Gameflow.java
package Console;

public class Gameflow {
    private static final int NUM_PLAYERS = 2;
    private static final int SIZE = 8;
    private Board board;
    private Player[] players;
    private Logic logic;
    private ConsoleUI display;
    private int validTurns = 0, currentPlayer = 0;

    /**
     * Constructor for Gameflow.
     */
    public Gameflow() {
        ConsoleGameInfo gameInfo = ConsoleGameInfo.getInstance();
        this.board = gameInfo.getBoard();
        this.players = gameInfo.getPlayers();
        this.logic = gameInfo.getLogic();
        this.display = new ConsoleUI();
    }

    /**
     * Runs the games loop.
     */
    public void run() {
        System.out.println("First player is black");
        while (!shouldStop()) {
            playOneTurn();
            this.currentPlayer = (this.currentPlayer + 1) % NUM_PLAYERS;
        }
        //ANNOUNCE WINNER
        System.out.println("No more moves to play");
        this.display.printBoard(this.board);
        int winner = this.logic.getWinner(this.players);
        System.out.println("Scoreboard:");
        System.out.println(this.players[0] + "\t\t\t" + this.players[0].getScore());
        System.out.println(this.players[1] + "\t\t\t" + this.players[1].getScore());
        if (this.players[0].getScore() == this.players[1].getScore())
            System.out.println("It's a tie!");
        else System.out.println("Winner is '" + this.players[winner]
                + "' with score of: " + this.players[winner].getScore());
    }

    /**
     * Plays one move of the game.
     */
    public void playOneTurn() {
        System.out.println("Current board:");
        System.out.println("X: " + this.players[0].getScore() + "\t" + "O" + this.players[1].getScore());
        this.display.printBoard(this.board);
        int temp = this.validTurns;
        this.validTurns += this.players[this.currentPlayer].makeMove(this.board, this.logic);
        if (this.validTurns == temp) {
            this.validTurns = 0;
        }
    }

    /**
     * Checks if the game should end.
     *
     * @return true if board is full or no more moves for a player, false otherwise.
     */
    public boolean shouldStop() {
        return this.board.isFull() || this.validTurns == NUM_PLAYERS;
    }
}